python/dex/interop/jax.py

```diff
@@ -271,5 +271,127 @@ def tuple_string(prefix):
+_is_linear_input = jax.ad.is_undefined_primal
+
+
+def dex_call_evaluate_linearized_transpose(cotangents, *args, func_atom):
+  """Evaluates the transpose of a function atom.  """
+
+  # `func_atom` is assumed to be of the form of `evaluate_linearized` from
+  # `dex_call_jvp`, applied to a some function atom, called `f`, say.
+  # Concretely, if `f` has three primal arguments, `func_atom` should look like:
+  # ```
+  # \ x0 x1 x2 u0 u1 u2.
+  #   intermediate_linearized = linearize f (x0, x1, x2)
+  #   snd intermediate_linearized (u0 u1 u2)
+  # ```
+  # In particular, its arguments are assumed to be `num_primals` primal inputs,
+  # followed by `num_primals` tangent inputs.
+
+  assert len(args) % 2 == 0
+  num_primals = len(args) // 2
+
+  primals, tangents = args[:num_primals], args[num_primals:]
+
+  # Helper functions to build strings of primal and tangent inputs.
+  def arg_string(prefix, index_set):
+    return " ".join(f"{prefix}{i}" for i in index_set)
+
+  def tuple_string(prefix, index_set):
+    return "(" + ", ".join(f"{prefix}{i}" for i in index_set) + ")"
+
+  # JAX uses `UndefinedPrimal` instances to mark input variables which the
+  # function needs to be transposed with respect to, and (consequently) for
+  # which no concrete values are available. `_is_linear_input` tests if the
+  # input is such an instance.
+  #
+  # `func_atom` is only guaranteed to be linear in its tangent inputs, so we
+  # check here that we're not expected to tranpose it with respect to any primal
+  # inputs. JAX *should* take care of this automatically, but this mechanism is
+  # somewhat poorly documented so its worth double checking.
+  if any(_is_linear_input(p) for p in primals):
+    raise RuntimeError("Primal inputs to transpose primitive are undefined.")
+
+  # Add `func_atom` as a variable `linearized` in the context.
+  env = api.insert(func_atom.module, api.as_cstr("linearized"), func_atom)
+
+  # Form lists of the indices in `tangents` which correspond to linear inputs
+  # (which we are expected to transpose w.r.t.) and constant inputs (which we
+  # are not).
+  tangent_input_indices = [
+      i for i, t in enumerate(tangents) if _is_linear_input(t)
+  ]
+  tangent_constant_indices = [
+      i for i, t in enumerate(tangents) if not _is_linear_input(t)
+  ]
+
+  # In this case, there are no cotangents to output. Not sure if JAX would skip
+  # calling this function in this case or not.
+  if len(tangent_input_indices) == 0:
+    return (None,) * len(args)
+
+  # Form a lambda which partially evaluates `linearized` at the constant primal
+  # and constant tangent values, with the remaining arguments (i.e. the linear
+  # input tangents) combined into a tuple, and then transpose the lambda.
+  #
+  # For a three-input primal function with constant input for the tangent
+  # parameter at index 1, the evaluated string should look like:
+  # ```
+  # \ x0 x1 x2 u1 ct.
+  #   transposeLinear (\(t0, t2). linearized x0 x1 x2 t0 u1 t2) ct
```

Comment:
This works, although one downside of this trick is that it effectively makes all Dex primitives do checkpointing as if they were decorated with `jax.remat`.

Reply:
Right, I see. It seems to me that in order to avoid this we would need to do the closure trick you talked about in the previous PR, that is to extract all intermediate arrays from the computation of `linearized` in the JVP op in order to re-use them here. I'm not sure how we would extract these intermediate arrays with the current version of the Dex foreign API.

Is there a way we could do this? Or am I missing something and is there another way to avoid the checkpointing behaviour?

Reply:
Yes, this improvement would require that API and I don't think that we can emulate it in any way.